src/storage/src/storage_state/async_storage_worker.rs

```diff
@@ -0,0 +1,166 @@
+//!
+//! CAUTION: This is not meant for high-throughput data processing but for
+//! one-off requests that we need to do every now and then.
+
+use std::marker::PhantomData;
+use std::sync::Arc;
+
+use differential_dataflow::lattice::Lattice;
+use timely::progress::{Antichain, Timestamp};
+use tokio::sync::mpsc;
+use tokio::sync::Mutex;
+use tracing::Instrument;
+
+use mz_persist_client::cache::PersistClientCache;
+use mz_persist_types::Codec64;
+use mz_repr::GlobalId;
+use mz_service::local::Activatable;
+use mz_storage_client::controller::CollectionMetadata;
+use mz_storage_client::controller::ResumptionFrontierCalculator;
+use mz_storage_client::types::sources::IngestionDescription;
+
+/// A worker that can execute command that come in on a channel and returns
+/// responses on another channel. This is useful in places where we can't
+/// normally run async code, such as the timely main loop.
+#[derive(Debug)]
+pub struct AsyncStorageWorker<T: Timestamp + Lattice + Codec64> {
+    tx: mpsc::UnboundedSender<(tracing::Span, AsyncStorageWorkerCommand<T>)>,
+    rx: crossbeam_channel::Receiver<AsyncStorageWorkerResponse<T>>,
+}
+
+/// Commands for [AsyncStorageWorker].
+#[derive(Debug)]
+pub enum AsyncStorageWorkerCommand<T: Timestamp + Lattice + Codec64> {
+    /// Calculate a recent resumption frontier for the ingestion.
+    CalculateResumeFrontier(
+        GlobalId,
+        IngestionDescription<CollectionMetadata>,
+        PhantomData<T>,
+    ),
+}
+
+/// Responses from [AsyncStorageWorker].
+#[derive(Debug)]
+pub enum AsyncStorageWorkerResponse<T: Timestamp + Lattice + Codec64> {
+    /// An `IngestionDescription` with a calculated, recent resume upper.
+    IngestDescriptionWithResumeUpper(
+        GlobalId,
+        IngestionDescription<CollectionMetadata>,
+        Antichain<T>,
+    ),
+}
+
+impl<T: Timestamp + Lattice + Codec64> AsyncStorageWorker<T> {
+    /// Creates a new [`AsyncStorageWorker`].
+    ///
+    /// IMPORTANT: The passed in `activatable` is activated when new responses
+    /// are added the response channel. It is important to not sleep the thread
+    /// that is reading from this via [`try_recv`](Self::try_recv) when
+    /// [`is_empty`](Self::is_empty) has returned `false`.
+    pub fn new<A: Activatable + Send + 'static>(
+        activatable: A,
+        persist_clients: Arc<Mutex<PersistClientCache>>,
+    ) -> Self {
+        let (command_tx, mut command_rx) = mpsc::unbounded_channel::<(tracing::Span, _)>();
+        let (response_tx, response_rx) = crossbeam_channel::unbounded();
+
+        mz_ore::task::spawn(|| "AsyncStorageWorker", async move {
+            while let Some(cmd) = command_rx.recv().await {
+                // Peel off all available commands. This allows us to catch up
+                // if we fall behind.
+                let mut commands = vec![cmd];
+                while let Ok(cmd) = command_rx.try_recv() {
+                    commands.push(cmd);
+                }
```

Comment:
Can you explain the benefit of collecting the currently pending commands in a vec here vs. reading them one-by-one in the above while loop?